message/messaging_service.hh

```diff
@@ -260,6 +260,7 @@ public:
         rack,
         dc,
         all,
+        transitional, // encrypt all outgoing, but do not enforce incoming.
```

Comment:
Seems to me that transitional is orthogonal to none/rack/dc/all and so should be a separate config item.

encrypt_what: which connections should be encrypted
transitional: whether we ignore mismatches in encrypt_what

Reply:
Not true. Transitional also forces all outgoing connections to be encrypted. I.e. we encrypt outgoint, but accept unencrypted.

Reply:
It does, but should it? It makes the transition look like none -> all -> dc. Why? Shouldn't we transition none->dc?

Reply:
1.) dc and rack should be deprecated. I know you want to avoid TLS overhead for local networks etc, but the recent issues with topology not propagating properly and breaking TLS-sct should be enough of an argument that these setups are both brittle as well as potentially insecure. 
2.) We cannot go none -> dc (or rack) without downtime (i.e. restarting all nodes at the same time) since they otherwise will disagree on accepted communication mode. By going this path, we ensure we can do it node-by-node.

Reply:
I don't see why dc/rack TLS control should be deprecated. dc/rack are stable concepts and it's not like a node can change its dc/rack association at will.

Reply:
Not dc and rack as concepts, but encryption limited to them. Mainly because it is a chicken-egg case, since we need to do interop to know rack/dc relation, and we've seen issues with it breaking when this is not propagated properly. 
But allowing some network traffic to be unencrypted is also a potential security hole if a user fuddles his private network setup in some way.

Reply:
We allow non-TLS connections.

I think it's reasonable to allow connections within the datacenter to be unencrypted and only encrypt them across datacenters.

Didn't we have a mechanism where the initial connection was encrypted, and afterwards we used unencrypted connections if we were in the same rack/dc? Or maybe I'm mixing it with compression.

Reply:
No, we do. But this only affects outgoing connection. If a node A thinks it is known to another node B, and uses non-TLS, assuming the rules should allow this, but the info known in the B target does not match A:s view of itself, we will fail. Which is apparently what happened in the raft-tls issue. 

While I agree about the in-datacenter stuff in principle, my objections are based on a.) the fail(s) like above, and b.) people not securing their "data center" properly - i.e. nw. breaches.

Reply:
If someone chooses not to use TLS, that's on them. I don't think we should enforce it.

Reply:
"enforce" here meaning "apply dc/rack-like rules". In which case we _do_ (and must) enforce TLS iff node (not) belongs to relevant topology group.